Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,185 @@
+  | `($(_) $n:num $i:num) => `($i:num#$n:num)
+  | _                => throw ()
+
+instance : Repr (BitVec n) where
+  reprPrec a _ := repr a.val ++ "#" ++ repr n
+
+instance : ToString (BitVec n) where
+  toString a := toString (repr a)
+
+/-- Theorem for normalizing the bit vector literal representation. -/
+@[simp] theorem ofNat_eq_ofNat : @OfNat.ofNat (BitVec n) i _ = BitVec.ofNat n i := rfl
+
+/-- Addition for bit vectors. -/
+protected def add (x y : BitVec n) : BitVec n :=
+  { val := x.val + y.val }
+/-- Subtraction for bit vectors. -/
+protected def sub (x y : BitVec n) : BitVec n :=
+  { val := x.val - y.val }
+/-- Multiplication for bit vectors. -/
+protected def mul (x y : BitVec n) : BitVec n :=
+  { val := x.val * y.val }
+/-- Modulo for bit vectors. -/
+protected def mod (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.val.isLt⟩ }
+/-- Division for bit vectors. -/
+protected def div (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val / y.val, Nat.lt_of_le_of_lt (Nat.div_le_self _ _) x.val.isLt⟩ }
+/-- Less than for bit vectors. -/
+protected def lt (x y : BitVec n) : Bool :=
+  x.val < y.val
+/-- Less than or equal to for bit vectors. -/
+protected def le (x y : BitVec n) : Bool :=
+  x.val ≤ y.val
+
+instance : Add (BitVec n) := ⟨BitVec.add⟩
+instance : Sub (BitVec n) := ⟨BitVec.sub⟩
+instance : Mul (BitVec n) := ⟨BitVec.mul⟩
+instance : Mod (BitVec n) := ⟨BitVec.mod⟩
+instance : Div (BitVec n) := ⟨BitVec.div⟩
+instance : LT (BitVec n)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec n)  := ⟨fun x y => BitVec.le x y⟩
+
+/-- Bitwise and for bit vectors. -/
+protected def and (x y : BitVec n) : BitVec n :=
+  { val := x.val &&& y.val }
+/-- Bitwise or for bit vectors. -/
+protected def or (x y : BitVec n) : BitVec n :=
+  { val := x.val ||| y.val }
+/-- Bitwise xor for bit vectors. -/
+protected def xor (x y : BitVec n) : BitVec n :=
+  { val := x.val ^^^ y.val }
+/-- Complement for bit vectors. -/
+protected def complement (x : BitVec n) : BitVec n :=
+  0 - (x + .ofNat n 1)
+/-- Shift left for bit vectors. -/
+protected def shiftLeft (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat <<< s)
+/-- Shift right for bit vectors. -/
+protected def shiftRight (a : BitVec n) (s : Nat) : BitVec n :=
+  .ofNat n (a.toNat >>> s)
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Rotate left. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Rotate right. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Append. -/
+def append (a : BitVec n) (b : BitVec m) : BitVec (n+m) :=
```

Comment:
As I commented on the mathlib PR, this would be clearer with the arguments renamed as `append msb lsb`.

Reply:
I will do that. Thanks.